feat(ownership): Automatic assignment of owners

georgepsarakis opened the pull request:
Enables configurable automatic assignment of new events to owners, according to the Ownership Rules, instead of only suggesting.

- [x] Add new boolean field in `ProjectOwnership` table: `auto_assignment`.
- [x] Include new field in the corresponding serializer.
- [x] Modify endpoint to update flag value, if included in the `PUT` request: `/projects/<organization_slug>)/<project_slug>/ownership`
- [x] UI: Add JSX form in `Issue Owners` settings page, in order to allow users to modify the new boolean flag.
- [x] Update `EventManager` flow to auto-assign the `Group` of the new event according to the matched rules (if any), unless it has already been assigned.
- [x] Add tests.

Closes #10908 .

markstory commented on src/sentry/api/endpoints/project_ownership.py:
```diff
@@ -79,6 +82,21 @@ def save(self):
+
+        logger = logging.getLogger(__name__)
+        changed = False
+        logger.info(self.object)
```
Why are you logging the entire object?

georgepsarakis replied:
This is temporary code, used for debugging. I intended to remove it for the final review.

markstory commented on src/sentry/event_manager.py:
```diff
@@ -1136,6 +1136,30 @@ def _save_aggregate(self, event, hashes, release, **kwargs):
+            from sentry.models import GroupAssignee, ProjectOwnership, Team, User
+            ownership = ProjectOwnership.objects.filter(
+                            project_id=project.id
+                        ).first()
```
You could use `ProjectOwnership.objects.get()` here.

georgepsarakis replied:
Since `get` will raise an [exception](https://docs.djangoproject.com/en/2.1/ref/models/querysets/#django.db.models.query.QuerySet.get) if it does not find an object with the given parameters and at this point we are not aware if the project has any ownership rules yet, I thought of using `filter` instead.

georgepsarakis replied:
Changed as suggested.

dcramer commented on src/sentry/event_manager.py:
```diff
@@ -1136,6 +1136,30 @@ def _save_aggregate(self, event, hashes, release, **kwargs):
+        # TODO: convert to model method
+        from sentry.models import GroupAssignee
+        assignee = GroupAssignee.objects.filter(group_id=group.id).first()
+        logger.info('-- starting auto-assignment process --')
```
log statements need to be thought of as singular events vs a text log of print statements

dcramer commented on src/sentry/api/endpoints/project_ownership.py:
```diff
@@ -79,6 +82,21 @@ def save(self):
+        changed = False
+        logger.info(self.object)
+
+        if 'autoAssignment' in self.object:
```
nit: its arguably simpler to just auto_assignment = self.object.get('autoAssignment') and then check if its set and does not match the current value

georgepsarakis replied:
Sounds indeed better, I will change this as suggested.

dcramer commented on src/sentry/event_manager.py:
```diff
@@ -1136,6 +1136,30 @@ def _save_aggregate(self, event, hashes, release, **kwargs):
+                logger.info('-- ownership {}'.format(ownership.id))
+                owners, _ = ProjectOwnership.get_owners(project.id, event.data)
+                if owners and owners is not ProjectOwnership.Everyone:
+                    # Try teams first
```
:+1:

dcramer commented on src/sentry/static/sentry/app/views/settings/project/projectOwnership/index.jsx:
```diff
@@ -114,6 +114,33 @@ class ProjectOwnership extends AsyncView {
         </Form>
+
+        <Form
+          apiEndpoint={`/projects/${organization.slug}/${project.slug}/ownership/`}
```
While it's probably ok to have this setting be specific in the ownership category, we generally have put options into ProjectOption in the past. I'll defer to the team here but having many ways to configure a project will create complexity and maintenance concerns down the road.

georgepsarakis replied:
Sure, just let me know which design you prefer and I will change it accordingly. I thought there was some parity with the `fallthrough` setting, but having it as an option in (I guess) `General Settings -> Event Settings` also makes perfect sense.

georgepsarakis commented:
@dcramer @markstory thanks a lot for the feedback! I will push code improvements and address comments shortly.

georgepsarakis commented:
- Simplified new `EventManager` methods.
- Added tests for API endpoint & `EventManager` functionality.

The endpoint tests will of course change if the design decision is made to use `ProjectOption` instead of the setting, but the automatic assignment actions are still valid, independent of the implementation.

In the meanwhile, I will also probably add the UI test, just to get a grasp of how JS tests work.

georgepsarakis commented:
I am sure you have other issues to attend to, but is there perhaps any further feedback on https://github.com/getsentry/sentry/pull/12580#discussion_r270121715 :smile:  ?

dcramer commented on src/sentry/event_manager.py:
```diff
@@ -1136,8 +1137,45 @@ def _save_aggregate(self, event, hashes, release, **kwargs):
                 event.event_id,
             )
 
+        self._handle_owner_assignment(project, group, event)
```
just calling this out for folks to eyeball: this is in the processing path for aggregates so performance might be a question

mattrobenolt commented on src/sentry/event_manager.py:
```diff
@@ -1136,8 +1137,45 @@ def _save_aggregate(self, event, hashes, release, **kwargs):
+        if not project.projectownership_set.exists():
+            return
+
+        ownership = project.projectownership_set.first()
```
Don't use `.first()`.

dcramer replied:
For context we try to avoid using .first() because it forces an order. It might be ok in this situation though as I think the result set is small.

If we're going to uniformly forbid it we should just do it at the queryset level.

georgepsarakis replied:
I thought about this too, but did not consider it an issue, because this table should be particularly small since it is bound by the number of projects, and the following query should always return a single row (unless I am mistaken):

```sql
EXPLAIN SELECT * FROM sentry_projectownership WHERE project_id = 2 ORDER BY id;
```

I can add an explicit ORDER BY clause: `order_by('project_id')` so that `.first()` does not attempt to sort with the primary key. Using the indexed column for sorting should be always optimal either way.

mattrobenolt commented on src/sentry/models/projectownership.py:
```diff
@@ -22,6 +22,7 @@ class ProjectOwnership(Model):
     raw = models.TextField(null=True)
     schema = JSONField(null=True)
     fallthrough = models.BooleanField(default=True)
+    auto_assignment = models.BooleanField(default=False)
```
We can't add this column without being nullable.

georgepsarakis replied:
Nice observation, I see your point, not being nullable it would break existing installations, right?

I think this should be instead handled by the migration though and the specification of a default column value, the default has not been set to `false` and the column is not nullable:
```
sentry=# \d+ sentry_projectownership
                                                          Table "public.sentry_projectownership"
     Column      |           Type           |                              Modifiers                               | Storage  | Stats target | Description 
-----------------+--------------------------+----------------------------------------------------------------------+----------+--------------+-------------
auto_assignment | boolean                  | not null                                                             | plain    |              | 
```

What I am suggesting is that the following `ALTER TABLE` statement should work with existing installations too:

```
ALTER TABLE sentry_projectownership ADD COLUMN new_boolean_col BOOLEAN DEFAULT false;
```

I will check again why this happens and whether it can actually be handled in the migration or not.

dcramer replied:
The issue actually isnt the default as much as it is performance. We can apply live migrations generally when columns are NULLable, as the lock is less restrictive. When doing a non-NULLable column, even one with a default, it has to take a (much) longer lasting lock.

georgepsarakis replied:
Thanks a lot for the detailed explanation @dcramer , I wasn't aware of this limitation!

georgepsarakis commented on src/sentry/event_manager.py:
```diff
@@ -1136,8 +1137,45 @@ def _save_aggregate(self, event, hashes, release, **kwargs):
+
+        owners, _ = ProjectOwnership.get_owners(project.id, event_data)
+        if owners and owners is not ProjectOwnership.Everyone:
+            # Try teams first, in alphabetical order by name
```
@markstory you mean not alter the order and give precedence to the team?

georgepsarakis replied:
Never mind, I just read https://github.com/getsentry/sentry/pull/12694 , seems great!

markstory commented:
Going to close this one as #12694 contains the work done here, and I'll be able to rebase/merge it more easily. Thank you for getting this going @georgepsarakis

georgepsarakis commented:
@markstory you are welcome. Glad if I could actually help, this is a very useful feature :smile: !
